djweb/dj_comp_hist/views.py

```diff
@@ -196,101 +198,83 @@ def browse(request):
-                                           Q(folder__box__number=boxes[1]))
-    elif len(boxes) == 1:
-        doc_objs = Document.objects.filter(folder__box__number=boxes[0])
+
+    # Sorted list of all available document types
+    doc_types = sorted(["minutes", "memo", "proposal", "letter", "receipt", "contract", "notice",
+                        "memo draft", "addendum", "change order", "form", "report", "invoice",
+                        "list", "routing sheet", "application", "note", "press release",
+                        "floor plan", "program", "pamphlet", "payroll sheet", "time record",
+                        "summary", "table", "telegram"])
+
+    # if no search_params, that means we're just loading the search page
+    if not request.GET.dict():
+        return render(request, 'advanced_search.jinja2', {'doc_types': doc_types})
     else:
-        doc_objs = Document.objects
-    # TODO: try to fix up below filter to be more accurate
-    try:
-        title = request.GET['title']
-        if title != "":
-            doc_objs = doc_objs.filter(Q(title__icontains=title))
-    except:
-        print("Error getting document title")
+        search_params, results = process_advanced_search_request(request)
+        search_objs = {
+            'results': results,  # = doc_objs
+            'search_params': search_params,
+            'doc_types': doc_types
+        }
+        return render(request, 'advanced_search.jinja2', search_objs)
 
-    try:
-        phrase = request.GET['contents']
-        if phrase != '':
-            raw_docs = Document.objects.raw(f'SELECT * from doc_fts WHERE text MATCH "{phrase}"')
+
+def process_advanced_search_request(request):
+    """
+    Processes one advanced search request and returns the search_results as a list of Document
+    objects as well as the search_params as a dict.
+
+    :param request:
+    :return:
+    """
+
+    search_params = request.GET.dict()
+    # doc_types param is a list but using request.GET['doc_type'] only returns the first one.
+    search_params['doc_types'] = request.GET.getlist('doc_type')
+
+    doc_objs = Document.objects
+
+    if search_params['title'] != '':
+        doc_objs = doc_objs.filter(Q(title__icontains=search_params['title']))
+
+    if search_params['text'] != '':
+        # allows quotation marks but only extracts the string in the middle
+        match = re.match('^[\'\"]?([a-zA-Z\d ]+)[\'\"]?$', search_params['text'])
+        if not match:
+            print(f"WARNING. Could not parse full text search string: {search_params['text']}.")
+        else:
+            raw_docs = Document.objects.raw(f'''SELECT * FROM doc_fts 
+                                                     WHERE text MATCH "{match.groups()[0]}";''')
             doc_ids = [doc.id for doc in raw_docs]
             doc_objs = doc_objs.filter(id__in=doc_ids)
 
-    except:
-        print("Error getting phrase")
+    if search_params['author'] != '':
```

Comment:
@srisi does this author filtering logic seem a bit off to you? Why are we only checking the first word in the query against all of the author fields?

Reply:
Agreed. I had copied an earlier version yesterday. However, on second thought, there is no reason why we shouldn't also use the case insensitive full text search for title, author, and recipient. -> updated.